pkg/core/resources/apis/meshservice/vip/allocator.go

```diff
@@ -0,0 +1,146 @@
+package vip
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/Nordix/simple-ipam/pkg/ipam"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/kumahq/kuma/pkg/core"
+	meshservice_api "github.com/kumahq/kuma/pkg/core/resources/apis/meshservice/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/pkg/core/user"
+	core_metrics "github.com/kumahq/kuma/pkg/metrics"
+)
+
+// Allocator manages IPs for MeshServices.
+// Each time allocator starts it initiates the IPAM based on existing MeshServices
+// We don't free addresses explicitly, but we always allocate next free IP to avoid a problem when we
+// 1) Remove Service A with IP X
+// 2) Add new Service B that gets IP X
+// 3) Clients that were sending the traffic to A now sends the traffic to B for brief amount of time
+// IPAM is kept in memory to avoid state management, so technically this problem can still happen when leader changes
+// However, leader should not change before TTL of a DNS that serves this VIP.
+//
+// It's technically possible to allocate all addresses by creating and removing services in the loop.
+// However, CIDR has range of 16M addresses, after that the component will just restart.
+type Allocator struct {
+	logger     logr.Logger
+	cidr       string
+	interval   time.Duration
+	metric     prometheus.Summary
+	resManager manager.ResourceManager
+}
+
+var _ component.Component = &Allocator{}
+
+func NewAllocator(
+	logger logr.Logger,
+	metrics core_metrics.Metrics,
+	resManager manager.ResourceManager,
+	cidr string,
+	interval time.Duration,
+) (*Allocator, error) {
+	metric := prometheus.NewSummary(prometheus.SummaryOpts{
+		Name:       "component_ms_vip_allocator",
+		Help:       "Summary of Inter CP Heartbeat component interval",
+		Objectives: core_metrics.DefaultObjectives,
+	})
+	if err := metrics.Register(metric); err != nil {
+		return nil, err
+	}
+
+	return &Allocator{
+		logger:     logger,
+		resManager: resManager,
+		cidr:       cidr,
+		interval:   interval,
+		metric:     metric,
+	}, nil
+}
+
+func (a *Allocator) Start(stop <-chan struct{}) error {
+	ticker := time.NewTicker(a.interval)
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+
+	kumaIpam, err := a.initIpam(ctx)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			start := core.Now()
```

Comment:
Why do we have this mocked time again? Can we avoid this unless we know we need it, though I don't think we ever actually need it?

Reply:
technically we don't need it. Removed.